### ISSUE TITLE
Ability to choose probability of each response when using random responses

### ISSUE THREAD

**Issue by maillouxc** [mockoon/mockoon#1016]
I have a need to simulate an error which occurs very infrequently. The random response feature is very useful, but I'd really like to be able to adjust the probability of each response.

**Comment by glaucoleme** [comment]
I think that this feature is related with [chooseResponse](https://github.com/mockoon/mockoon/blob/main/packages/commons-server/src/libs/response-rules-interpreter.ts#L43) function, between lines 49~53. [I did a draft](https://onecompiler.com/javascript/3z6x63y2q) just to try out a solution of how use the probability of each response could be implemented. 

But how this would work? Should an option be added at settings of each response to adjust this probability, which is enabled only when random response mode is activated? 

Also, the probability of all responses should sum up to 1 (100%), right? This seems a quite tricky, as more responses are added, all remaining responses probability are affected too. On other hand, without normalizing these scores, it might be confuse for user to handle it.

Just trying to contribuite into the discussion.

**Comment by maillouxc** [comment]
I think to get around the problem of having to make all the probabilities add up to 100 (which would be quite tricky to implement without a UI redesign) we could instead use weights. You would specify a weight value for each response. That is, a response with a weight of 2 should appear twice as often as a response with a weight of 1. Therefore percentage could be (responseWeight / sumOfWeights).

I think as far as how to specify it, a new small box would appear on the dropdown next to each response that can be sent where you can type the weight.

**Comment by glaucoleme** [comment]
> "Therefore percentage could be (responseWeight / sumOfWeights)." 

That is also what I aimed at the [draft](https://onecompiler.com/javascript/3z6x8b43p). Normalizing the values and then compare with random output. This code is badly made, as I did not have time to optimize it. But could you check if its propose fits what you thought?

**Comment by 255kb** [comment]
I'm not sure if many people would use this feature it seems quite specific. 
I don't want to add it to the UI and clutter it even more without having some popularity signal (i.e. issue upvotes). 

I could settle on adding this feature to the response settings:
![image](https://user-images.githubusercontent.com/7489814/235098179-4e2f3e07-396a-48d7-b6a4-161a25c539bd.png)

As per the implementation, I guess it would be simpler by adding weights to the responses like @maillouxc suggested:
- One input field per response, default to 1 (field could be empty to avoid any confusion)
- in the `chooseResponse` function we could create an array containing the responses UUIDs, and do the random on the length: 

Example: if we have two response, one weighted 1 (default) and the other 2, we would have this array: 
`['uuid1', 'uuid2', 'uuid2']` and keep the rest of the code nearly unchanged.

**Comment by 255kb** [comment]
> I do agree with the popularity signal, but how exactly does this works?

It's a bit off topic, but to answer your question it's not an exact science. Usually I look at the issue upvotes (👍), and other activities like comments. But I also listen to the users feedback and support requests. For example, the recent CRUD endpoints weren't subject of an open issue, but this was the biggest misconception in support requests. Lots of people expected it to work like that. So, I decided it was important. It's also a good feature making Mockoon a more serious competitor. 

I add to that a bit of "this would make sense", "this is a quick win and could be interesting", "this is highly upvoted but would take a lot of time", etc. 
A good example for the latter is the graphQL issue. The most upvoted, with no interaction whatsoever. I asked multiple times in the past for feedback as I don't work myself with graphQL and I got close to 0 replies. Also, this feature is wildly different from the REST approach and requires a lot of changes to the UI. 
I did a POC which is more or less working, but realized that it needed some persisting storage. That's why I worked on the data buckets first. 
It will probably come but I can only tackle one thing at a time. I try to do my best to find the right balance and also doing all the other stuff to make the project growth (tutorials) and sustainable (Pro plans). 

I hope this answers your request! 

I am also completely open to feedback, advice and help of any kind 😄

**Comment by InsaneBytecode** [comment]
May I work on this?

(This is technically my first contribution on this account, but I have another account that Is way more personal)

**Comment by InsaneBytecode** [comment]
Oh Nevermind, someone on another platform did something they where not meant to and my identity is leaked. I will be back on another account

**Comment by ramonparisg** [comment]
Hello! is there a workaround to achieve this? :)

**Comment by 255kb** [comment]
> Hello! is there a workaround to achieve this? :)

Not really. The only way I see is to duplicate responses. E.g. three 200 and one 500, will give you 3/4 200 and 1/4 500, more or less.

**Comment by 255kb** [comment]
I was revisiting this issue today and came up with an implementation idea that wouldn't require a model change or a major release.

TL;DR; make the random mode more realistic instead of being able to choose the probability. 

Two remarks/questions first:

- I think weighted responses would only make sense when using randomized mode.
- Is there real value in being able to precisely say “I want this response 2/3rd of the time,” or is the real issue that with two responses, the current randomization feels a bit too “dumb”? For example, with a 200 and a 500, you could get: 200, 200, 500, 500, 500, 500, 200, 200, 200… not very realistic.

Here's the implementation idea:

- No UI change.
- When random mode is enabled, responses would be weighted based on two criteria:
- Order in the list: the lower it is, the less frequently it’s served.
- Status code: for example, 4xx responses are served less often than 2xx, and 5xx less often than 4xx.

Curious to hear your thoughts on this approach.

**Comment by GigadatJRoss** [comment]
Just wanted to add another use case for this. We’re currently using the Mockoon CLI to simulate a 3rd party service for performance testing, where we’re often running millions of requests per test.

At this scale, getting a precise distribution is a bit of a challenge. Right now, we’re mimicking our production stats by essentially 'stacking' duplicate responses to get the ratios right. It works, but it's pretty tedious to maintain. If we need to update the ratio to reflect what we see in prod, we have to change or create dozens of identical responses just to keep the error rates accurate.

Having a more streamlined way to manage these distributions would be a great help for high-volume workflows like ours.

Hope this helps clarify some real-world use cases!

**Comment by 255kb** [comment]
One recent conversation with a user made me realize that simple weighted responses can be achieved using [custom rules](https://mockoon.com/docs/latest/route-responses/dynamic-rules/#1-target). Custom rules support templating in both the "target" and the "value" fields, you can implement logic like this:

<img width="1161" height="272" alt="Image" src="https://github.com/user-attachments/assets/d7f1c6cc-511e-4334-815e-df9d3fe8f285" />

or this: 

<img width="1160" height="291" alt="Image" src="https://github.com/user-attachments/assets/04d3c4f1-ea09-4672-9617-9a19ceebd694" />

Both examples result in a 20% chance of the 401 occurring (assuming a 200 default response and a 401 response with the rule). This method works with multiple responses, though the distribution ratio may not be perfectly precise the more responses you add. 

Another way to achieve more precise distribution is to generate a random number only once, in the first rule, store it in a global variable and check it in the other rules, it's definitely less practical but doable:

First response's rule: `{{setGlobalVar 'rand' (int 0 100)}}{{lt (getGlobalVar 'rand') 5}}`
Second response's rules: `{{gte (getGlobalVar 'rand') 5}}` AND `{{lt (getGlobalVar 'rand') 15}}`
Etc.
